### PR TITLE
Fix MuSGD head grouping for custom backbones

### PR DESCRIPTION
**Pull request by fcakyon**
MuSGD assumes the detection head is always layer 23. Custom model YAMLs can place it elsewhere, so their classification head silently misses the intended 3x learning rate.

Run this on `main`, then on this PR:

```python
from ultralytics.engine.trainer import BaseTrainer
from ultralytics.nn.tasks import DetectionModel

cfg = {
    "nc": 2,
    "backbone": [
        [-1, 1, "Conv", [16, 3, 2]],
        [-1, 1, "Conv", [32, 3, 2]],
        [-1, 1, "Conv", [64, 3, 2]],
    ],
    "head": [[[0, 1, 2], 1, "Detect", ["nc"]]],
}
model = DetectionModel(cfg, verbose=False)
trainer = object.__new__(BaseTrainer)
optimizer = trainer.build_optimizer(model, name="MuSGD", lr=0.01)
head = model.model[-1]
head_params = {id(p) for p in head.cv3.parameters()}
head_lrs = {
    group["lr"]
    for group in optimizer.param_groups
    if any(id(p) in head_params for p in group["params"])
}
print(f"Detect layer: {head.i}")
print(f"Classification head LR: {head_lrs}")
```

Current `main`:

```text
Detect layer: 3
Classification head LR: {0.01}
```

This PR:

```text
Detect layer: 3
Classification head LR: {0.03}
```

- for distillation, boosts the student head and not the teacher
- Deleted: fixed layer-23 regex and local `re` import

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 Improves MuSGD fine-tuning by reliably assigning higher learning rates to detection and semantic-segmentation head parameters.

### 📊 Key Changes
- Replaced regex-based parameter matching with direct inspection of the model’s final head.
- Automatically identifies parameters in the `cv3` and `one2one_cv3` layers for boosted learning rates.
- Preserves special handling for semantic-segmentation parameters such as `proto.semseg` and `SemanticSegment`.
- Supports wrapped models and student-model architectures through safer model unwrapping.
- Added the contributor @rudrakumar07 to the documentation author registry.

### 🎯 Purpose & Impact
- 🎯 Makes MuSGD fine-tuning more robust across supported model architectures.
- ⚡ Applies the higher learning rate to the intended head parameters without relying on fragile parameter-name patterns.
- 🧠 May improve adaptation speed and fine-tuning results for detection and segmentation models.
- ✅ Reduces the risk of missing relevant parameters when model structures or naming conventions change.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fcakyon, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms that your MuSGD optimizer update has been received; an engineer will assist with review. Please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean PR. The change replaces the brittle fixed-layer name match with parameter-identity selection from the unwrapped model’s student detection head, while preserving the semantic-segmentation handling and existing MuSGD parameter-group behavior.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged in #25532—thank you @fcakyon for making MuSGD fine-tuning more robust and reliable across detection and semantic-segmentation architectures!

By inspecting the model’s final head directly, this PR ensures higher learning rates reach the intended parameters while supporting wrapped and student-model designs. As Charles Darwin said, “It is not the strongest of the species that survives, nor the most intelligent, but the one most responsive to change.” This work helps Ultralytics models adapt more effectively as architectures evolve.

We also appreciate @rudrakumar07 for joining the documentation author registry. 🚀
